solidity/20241126-upgrade-path-testing.md

```diff
@@ -0,0 +1,119 @@
+
+The current foundry testing uses the following high level call flow.
+
+```mermaid
+graph LR
+    A["CommonTest.setUp()"] --> B["Setup.L1()"] --> C["Deploy.run()"]
+    C --> D["DeployImplementations.run()"]
+    C --> E["DeploySuperchain.run()"]
+    C --> F["OPCM.deploy()"]
+```
+
+This is roughly what each component does:
+
+- **`CommonTest.setUp()`:**
+  - Defines system config settings (ie. `useInterop`).
+  - provides reusable addresses (`alice` and `bob`)
+  - provides reusable generic contracts (`ERC20`)
+- **`Setup.L1()`:**
+  - Reads addresses from the deployments file stored on disk
+  - labels addresses to make traces more readable
+  - ie.
+    ```solidity
+      optimismPortal = IOptimismPortal(deploy.mustGetAddress("OptimismPortalProxy"));
+      vm.label(address(optimismPortal), "OptimismPortal");
+    ```
+- **`Deploy.run()`:**
+  - Deploys all necessary contracts via calls to `DeploySuperchain`, `DeployImplementations`.
+- **`DeploySuperchain.run()`:**
+  - Deploys superchain contracts (`SuperchainConfig` and `ProtocolVersions`).
+- **`DeployImplementations.run()`:**
+  - Deploys contracts (implementations and singletons) necessary for upgrading to the system on
+    `develop`.
+- **`OPCM.deploy()`:**
+  - Deploys all proxies and bespoke singleton contracts as necessary for a new OP Chain.
+
+This work would replace the `Deploy` script with a new `Upgrade` script, resulting in the following
+call flow:
+
+```mermaid
+graph LR
+    A["CommonTest.setUp()"] --> B["Setup.L1()"] -->  C["Upgrade.run()"]
+    C --> D["DeployImplementations.run()"]
+    C --> E["OPCM.upgrade()"]
+```
+
+A description of what new components would do is:
+
+- **`Upgrade.run()`:**
+  - Calls `op-deployer bootstrap superchain` to deploy new superchain contracts (`SuperchainConfig`
+    and `ProtocolVersions`), corresponding to the previous release.
+  - Calls `op-deployer bootstrap opcm` to deploy release OPCM, corresponding to the previous release.
+  - Calls to `DeployImplementations.run()` to deploy contracts necessary for upgrading to the system on `develop`.
+  - Parses the deployment output from `op-deployer` and writes to disk using the `Deploy.save()`
+    functions, so that `Setup.L1()` can read the deployment.
+- **`OPCM.upgrade()`:**
+  - Upgrades proxies to new implementation contracts and bespoke singleton contracts as necessary for a new OP Chain.
+  - This flow is descibed in detail in the [L1 Upgrades design](../protocol/l1-upgrades.md#release-process).
+
+This testing setup route would be indicated with a new `useUpgradedSystem` flag in `CommonTest`. The
+new flag could only be enabled when other flags (`useAltDAOverride`, `useLegacyContracts`,
+`useInteropOverride`, `customGasToken`) are disabled.
+
+Note that this testing would need to be run against an `anvil` node, as `op-deployer bootstrap`
+requires an RPC endpoint. I am not experienced with running `forge test` against an anvil node,
+so appreciate any gotchas I might be missing.
```

Comment:
Am I correct that the flow here would be for our forge tests to invoke `op-deployer` against an anvil RPC URL, then after op-deployer runs it would `vm.createSelectFork(anvilRpcUrl)`?

Reply:
> invoke `op-deployer` against an anvil RPC URL 

Yes. 

> then after op-deployer runs it would `vm.createSelectFork(anvilRpcUrl)`?

I had envisioned calling `forge test --fork-url anvilRpcUrl`. Creating the fork in the script seems less error prone so I like this, are there any other differences I should be aware of between those two approaches?

Reply:
> I had envisioned calling `forge test --fork-url anvilRpcUrl`. Creating the fork in the script seems less error prone so I like this

I am not certain but it's possible the `forge test --fork-url anvilRpcUrl` approach might not work because forge may always make requests at whatever the initial block is at fork time, i.e. I don't know if it follows the chain live, if that makes sense. So starting our fork from within the test/script itself after anvil is properly configured seems safer anyway

I don't think there should be any other differences. There are some `createSelectFork` things to be aware of if you switch between forks in the tests, like `vm.makePersistent`